src/main/cpp/util/path_windows.cc

```diff
@@ -275,104 +275,87 @@ void assignNUL(std::wstring* s) { s->assign(L"NUL"); }
 //   "D:\\Bar".
-template <typename char_type>
-static bool NormalizeWindowsPath(const std::basic_string<char_type>& path,
-                                 std::basic_string<char_type>* result) {
-  if (path.empty()) {
-    *result = path;
-    return true;
-  }
-  if (IsPathSeparator(path[0])) {
-    return false;
+template <typename C>
+std::basic_string<C> NormalizeWindowsPath(const std::basic_string<C>& p) {
+  if (p.empty()) {
+    return p;
   }
-
-  static const std::basic_string<char_type> kDot =
-      std::basic_string<char_type>(1, '.');
-  static const std::basic_string<char_type> kDotDot =
-      std::basic_string<char_type>(2, '.');
-
-  std::vector<std::basic_string<char_type> > segments;
-  std::basic_string<char_type>::size_type seg_start = path.size();
-  std::basic_string<char_type>::size_type total_len = 0;
-  for (std::basic_string<char_type>::size_type i =
-           HasUncPrefix(path.c_str()) ? 4 : 0;
-       i <= path.size(); ++i) {
-    if (i == path.size() || IsPathSeparator(path[i])) {
-      // The current character ends a segment.
-      if (seg_start < path.size() && i > seg_start) {
-        std::basic_string<char_type> seg =
-            i == path.size() ? path.substr(seg_start)
-                             : path.substr(seg_start, i - seg_start);
-        if (seg == kDotDot) {
-          if (segments.empty() || segments.back() == kDotDot) {
-            // Preserve ".." if the path is relative and there are only ".."
-            // segment(s) at the front.
-            segments.push_back(seg);
-            total_len += 2;
-          } else if (segments.size() == 1 && segments.back() == kDot) {
-            // Replace the existing "." if that was the only path segment.
-            segments[0] = seg;
-            total_len = 2;
-          } else if (segments.size() > 1 ||
-                     !HasDriveSpecifierPrefix(segments.back().c_str())) {
-            // Remove the last segment unless the path is already at the root
-            // directory.
-            total_len -= segments.back().size();
-            segments.pop_back();
-          }
-        } else if (seg == kDot) {
-          if (segments.empty()) {
-            // Preserve "." if and only if it's the first path segment.
-            // Subsequent segments may replace this segment.
-            segments.push_back(seg);
-            total_len = 1;
-          }
+  typedef std::basic_string<C> Str;
+  static const Str kDot(1, '.');
+  static const Str kDotDot(2, '.');
+  std::vector<std::pair<Str::size_type, Str::size_type> > segments;
+  Str::size_type seg_start = Str::npos;
+  bool first = true;
+  bool abs = false;
+  bool starts_with_dot = false;
+  for (Str::size_type i = HasUncPrefix(p.c_str()) ? 4 : 0; i <= p.size(); ++i) {
+    if (seg_start == Str::npos) {
+      if (i < p.size() && p[i] != '/' && p[i] != '\\') {
+        seg_start = i;
+      }
+    } else {
+      if (i == p.size() || (p[i] == '/' || p[i] == '\\')) {
```

Comment:
Just `i == p.size() || p[i] == '/' || p[i] == '\\`

Reply:
Nice, thanks! Will do in the imported PR.